modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/index/AbstractRebuildIndexTest.java

```diff
@@ -0,0 +1,236 @@
+import org.apache.ignite.failure.StopNodeFailureHandler;
+import org.apache.ignite.internal.IgniteEx;
+import org.apache.ignite.internal.IgniteInternalFuture;
+import org.apache.ignite.internal.processors.cache.CacheMetricsImpl;
+import org.apache.ignite.internal.processors.cache.GridCacheContext;
+import org.apache.ignite.internal.processors.cache.index.IndexesRebuildTaskEx.BreakRebuildIndexConsumer;
+import org.apache.ignite.internal.processors.cache.index.IndexesRebuildTaskEx.StopRebuildIndexConsumer;
+import org.apache.ignite.internal.processors.cache.persistence.CacheDataRow;
+import org.apache.ignite.internal.util.lang.IgniteThrowableBiPredicate;
+import org.apache.ignite.internal.util.typedef.G;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+import org.jetbrains.annotations.Nullable;
+
+import static java.util.stream.Collectors.toList;
+import static org.apache.ignite.cluster.ClusterState.ACTIVE;
+import static org.apache.ignite.internal.processors.cache.index.IndexesRebuildTaskEx.addCacheRowConsumer;
+import static org.apache.ignite.internal.processors.cache.index.IndexesRebuildTaskEx.nodeName;
+import static org.apache.ignite.testframework.GridTestUtils.deleteIndexBin;
+
+/**
+ * Base class for testing index rebuilds.
+ */
+public abstract class AbstractRebuildIndexTest extends GridCommonAbstractTest {
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        IndexesRebuildTaskEx.clean(getTestIgniteInstanceName());
+
+        stopAllGrids();
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        super.afterTest();
+
+        IndexesRebuildTaskEx.clean(getTestIgniteInstanceName());
+
+        stopAllGrids();
+        cleanPersistenceDir();
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String igniteInstanceName) throws Exception {
+        return super.getConfiguration(igniteInstanceName)
+            .setConsistentId(igniteInstanceName)
+            .setFailureHandler(new StopNodeFailureHandler())
+            .setDataStorageConfiguration(
+                new DataStorageConfiguration()
+                    .setDefaultDataRegionConfiguration(new DataRegionConfiguration().setPersistenceEnabled(true))
+            ).setCacheConfiguration(cacheCfg(DEFAULT_CACHE_NAME, null));
+    }
+
+    /** {@inheritDoc} */
+    @Override protected IgniteEx startGrid(int idx) throws Exception {
+        IgniteEx n = super.startGrid(idx);
+
+        n.cluster().state(ACTIVE);
+
+        return n;
+    }
+
+    /**
+     * Prepare cluster for test.
+     *
+     * @param keys Key count.
+     * @return Coordinator.
+     * @throws Exception If failed.
+     */
+    protected IgniteEx prepareCluster(int keys) throws Exception {
+        IgniteEx n = startGrid(0);
+
+        if (n.cluster().state() != ACTIVE)
```

Comment:
extra operation i hope

Reply:
The operation is needed if we start the nodes at the beginning, and then activate them, for example, here:
**ResumeRebuildIndexTest#checkRestartRebuildIndexes**

Reply:
But `checkRestartRebuildIndexes` calls `startGrid(int)` which already activate the cluster and `ResumeRebuildIndexTest` do NOT override `startGrid(int)`.

I think one line economy having a `prepareCluster(int)` instead of `startGrid(int); populateCache(int)` in tests saves nothing, but makes test code uncertain.